client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java

```diff
@@ -201,13 +202,21 @@ public Configuration getHadoopConf() {
         RssShuffleDataIterator iterator = new RssShuffleDataIterator<K, C>(
             shuffleDependency.serializer(), shuffleReadClient,
             readMetrics);
-        iterators.add(iterator);
+        CompletionIterator<Product2<K, C>, RssShuffleDataIterator<K, C>> completionIterator =
```

Comment:
If we use AQE, we could have many iterators, if we don't release the memory of them after we use them, we may occur OOM.

Reply:
`org.apache.spark.util.CompletionIterator` will clean up as soon as its wrapped `RssShuffleDataIterator.hasNext` returns `false `:

```
// org.apache.spark.util.CompletionIterator
  def hasNext: Boolean = {
    val r = iter.hasNext  // iter => RssShuffleDataIterator
    if (!r && !completed) {
      completed = true
      // reassign to release resources of highly resource consuming iterators early
      iter = Iterator.empty.asInstanceOf[I]
      completion()
    }
    r
  }

  def completion(): Unit  // completion() => RssShuffleDataIterator.cleanup
```

After this PR, if there is no special case for the `Spark Task`, the timing of resource cleanup is still when the `RssShuffleDataIterator` ends, not when the `Spark Task` ends.

This is the same behavior as before the PR.

Reply:
> `org.apache.spark.util.CompletionIterator` will clean up as soon as its wrapped `RssShuffleDataIterator.hasNext` returns `false `:
> 
> ```
> // org.apache.spark.util.CompletionIterator
>   def hasNext: Boolean = {
>     val r = iter.hasNext  // iter => RssShuffleDataIterator
>     if (!r && !completed) {
>       completed = true
>       // reassign to release resources of highly resource consuming iterators early
>       iter = Iterator.empty.asInstanceOf[I]
>       completion()
>     }
>     r
>   }
> 
>   def completion(): Unit  // completion() => RssShuffleDataIterator.cleanup
> ```
> 
> After this PR, if there is no special case for the `Spark Task`, the timing of resource cleanup is still when the `RssShuffleDataIterator` ends, not when the `Spark Task` ends.
> 
> This is the same behavior as before the PR.

OK, I got it. Good catch.